In [1]:
import os
import re
import pathlib
import pandas as pd
from datetime import datetime, timedelta
# from pdf_to_txt import extract_using_pdfminer

## extract_using_pdfminer() working and non-working examples

In [2]:
p = "/nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/100400.pdf"
output_txt = "/nrcan_p2/data/02_intermediate/"
df, failed_pages = extract_using_pdfminer(p, output_txt, N_PAGES=None, WRITE_TXT=False)

/nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/100400.pdf


5it [00:01,  3.70it/s]

('Rotated page found:', 4, 39, 36, 1)
('Rotated page found:', 4, 39, 36, 1)
Rotated page still returns nothing: /nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/100400.pdf -- Pg No.4


6it [00:01,  3.77it/s]

('Rotated page found:', 5, 25, 22, 1)
('Rotated page found:', 5, 25, 22, 1)
Rotated page still returns nothing: /nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/100400.pdf -- Pg No.5


8it [00:01,  3.54it/s]

('Rotated page found:', 7, 30, 25, 3)
('Rotated page found:', 7, 30, 25, 3)
Rotated page still returns nothing: /nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/100400.pdf -- Pg No.7
('Rotated page found:', 8, 20, 17, 1)
('Rotated page found:', 8, 20, 17, 1)
Rotated page still returns nothing: /nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/100400.pdf -- Pg No.8


17it [00:03,  4.87it/s]


In [3]:
df

obj_type  pg      pos_x0  \
0    <class 'pdfminer.layout.LTTextBoxHorizontal'>   0   98.730000   
1    <class 'pdfminer.layout.LTTextBoxHorizontal'>   0  153.410000   
2    <class 'pdfminer.layout.LTTextBoxHorizontal'>   0  237.870000   
3    <class 'pdfminer.layout.LTTextBoxHorizontal'>   0  177.140000   
4    <class 'pdfminer.layout.LTTextBoxHorizontal'>   0  250.040000   
..                                             ...  ..         ...   
272  <class 'pdfminer.layout.LTTextBoxHorizontal'>  16  323.590000   
273  <class 'pdfminer.layout.LTTextBoxHorizontal'>  16  368.230000   
274  <class 'pdfminer.layout.LTTextBoxHorizontal'>  16  412.380000   
275             <class 'pdfminer.layout.LTFigure'>  16   -1.469833   
276             <class 'pdfminer.layout.LTFigure'>  16   -1.469833   

         pos_y0      pos_x1      pos_y1  \
0    658.029300  473.863100  669.704900   
1    633.319300  416.201699  644.994900   
2    583.389300  324.476396  595.064900   
3    558.919300  384.814591  570.594900   
4    485.203400  317.214523  496.140200   
..          ...         ...         ...   
272  131.051500  350.645200  383.454200   
273  131.290200  393.025937  377.700600   
274  131.291500  440.312200  378.449500   
275   -1.121124  578.429810  755.921112   
276   -1.121124  578.429810  755.921112   

                                                  text  
0    GPAVI TY  AND  MAGNETIC  DATA  AT  THE  MI D- ...  
1    HUD SON  65- 06 ,  HUDSOi'J 66-1 9  AND  HUDSO...  
2                                  DI GITIZED  DATA \n  
3         RELEASED  THRO UGH  THE  GSC. OPEN  Fl LE \n  
4                                      PREPARED  BY \n  
..                                                 ...  
272  a \n0 \n0 \n12 \n-7 \n-27 \n71 \n1 \n11 \n-45 ...  
273  0 \n0 \n0 \n3J \n4 \n25 \n4 \n8 \n0 \n-8 \n-40...  
274  0 \n0 \nc \n2 \n29 \n12 \n-13 \ni) \n29 \n11 \...  
275                                               None  
276                                               None  

[277 rows x 7 columns]

In [4]:
failed_pages

file  ipage  \
0  /nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf...      4   
1  /nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf...      5   
2  /nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf...      7   
3  /nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf...      8   

                                                page  
0  <PDFPage: Resources={'Font': {'C0_0': <PDFObjR...  
1  <PDFPage: Resources={'Font': {'C0_0': <PDFObjR...  
2  <PDFPage: Resources={'Font': {'C0_0': <PDFObjR...  
3  <PDFPage: Resources={'Font': {'T1_0': <PDFObjR...

In [5]:
p = "/nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/102416.pdf"
output_txt = "/nrcan_p2/data/02_intermediate/"
df, failed_pages = extract_using_pdfminer(p, output_txt, N_PAGES=None, WRITE_TXT=False)

0it [00:00, ?it/s]

/nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/102416.pdf


46it [00:05,  7.79it/s]


In [6]:
df

obj_type  pg    pos_x0    pos_y0  \
0    <class 'pdfminer.layout.LTTextBoxHorizontal'>   0  276.4200  669.5428   
1    <class 'pdfminer.layout.LTTextBoxHorizontal'>   0  261.9000  657.5428   
2    <class 'pdfminer.layout.LTTextBoxHorizontal'>   0  267.7000  637.8628   
3    <class 'pdfminer.layout.LTTextBoxHorizontal'>   0  284.9200  613.8628   
4    <class 'pdfminer.layout.LTTextBoxHorizontal'>   0  293.0100  543.4869   
..                                             ...  ..       ...       ...   
857  <class 'pdfminer.layout.LTTextBoxHorizontal'>  45   32.3600  144.5338   
858  <class 'pdfminer.layout.LTTextBoxHorizontal'>  45   49.3490  136.8626   
859  <class 'pdfminer.layout.LTTextBoxHorizontal'>  45   32.1144  122.4502   
860  <class 'pdfminer.layout.LTTextBoxHorizontal'>  45   48.8020  114.3010   
861             <class 'pdfminer.layout.LTFigure'>  45    0.0000    0.0000   

         pos_x1      pos_y1                                               text  
0    420.055200  680.966800                      This document was produced \n  
1    435.701200  668.966800         by  scanning the original  publication. \n  
2    428.873920  649.286800               Ce document est Ie  produit d'une \n  
3    414.939360  637.286800  numerisation par balayage \nde la publication ...  
4    388.734374  558.861700                                   BULLETIN  200 \n  
..          ...         ...                                                ...  
857  429.039000  154.131400  195  Petrology and structure of Thor-Odin gnei...  
858  192.784760  146.461400      and  J.  M.  Moore,  Jr.,  S2.50 (M42·195) \n  
859  430.091600  132.051400  200  Part  I- Biostratigrttphy of some  Ea rly...  
860  378.925900  123.901400  Silurian Arthropoda from  Arctic and eastern C...  
861  454.080002  680.640106                                               None  

[862 rows x 7 columns]

In [7]:
df["text"][0]

'This document was produced \n'

## Unprocessed pdfs

In [2]:
! ls /nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf | wc -l

7183


In [3]:
! find /nrcan_p2/data/02_intermediate/20201006/geoscan/pdf/v1_20201125 -name "*.pdfminer_split.csv" -type f | wc -l

6907


In [4]:
! find /nrcan_p2/data/02_intermediate/20201006/geoscan/pdf/v1_20201125 -name "*.pdfminer_split_failed_pages.csv" -type f | wc -l

9


In [5]:
! wc -l /nrcan_p2/data/02_intermediate/20201006/geoscan/pdf/v1_20201125/timeout_files.txt

273 /nrcan_p2/data/02_intermediate/20201006/geoscan/pdf/v1_20201125/timeout_files.txt


In [6]:
! cat /nrcan_p2/data/02_intermediate/20201006/geoscan/pdf/v1_20201125/empty_files.txt

/nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/247715.pdf
/nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/247715.pdf
/nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/247715.pdf


In [7]:
def get_pdf_ids(path, pattern):
    pdfs = []
    for (dirpath, dirnames, filenames) in os.walk(path):
        for filename in filenames:
            if filename.endswith(pattern):
                pdfs.append(filename.replace(pattern, ""))
    return pdfs

In [8]:
pdfs = get_pdf_ids("/nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf", ".pdf")

In [9]:
len(pdfs)

7183

In [10]:
processed_pdfs = get_pdf_ids("/nrcan_p2/data/02_intermediate/20201006/geoscan/pdf/v1_20201125", ".pdfminer_split.csv")

In [11]:
len(processed_pdfs)

6907

In [12]:
unprocessed_pdfs = set(pdfs) - set(processed_pdfs)

In [13]:
len(unprocessed_pdfs)

276

In [14]:
unprocessed_pdfs

{'211796',
 '211866',
 '211887',
 '247715',
 '247749',
 '247750',
 '247751',
 '247752',
 '247753',
 '247754',
 '247758',
 '247834',
 '247853',
 '247867',
 '248213',
 '248226',
 '248231',
 '248252',
 '261482',
 '261487',
 '261488',
 '261500',
 '261570',
 '262739',
 '262740',
 '262741',
 '285377',
 '285402',
 '285475',
 '285518',
 '285524',
 '285568',
 '285600',
 '286059',
 '286193',
 '286243',
 '286261',
 '286263',
 '286264',
 '286265',
 '286319',
 '286334',
 '287145',
 '287189',
 '287924',
 '287973',
 '288024',
 '288087',
 '288092',
 '288642',
 '288673',
 '288749',
 '288751',
 '288757',
 '288758',
 '288787',
 '288788',
 '288789',
 '288790',
 '288791',
 '288801',
 '288930',
 '288933',
 '288935',
 '289096',
 '289234',
 '289269',
 '289306',
 '289315',
 '289507',
 '289527',
 '289528',
 '289545',
 '289547',
 '289598',
 '289600',
 '291805',
 '291807',
 '291808',
 '291809',
 '291810',
 '291811',
 '291812',
 '291900',
 '291901',
 '291920',
 '292021',
 '292074',
 '292100',
 '292101',
 '292187',

In [15]:
pdfs_w_failed_pages = get_pdf_ids("/nrcan_p2/data/02_intermediate/20201006/geoscan/pdf/v1_20201125", ".pdfminer_split_failed_pages.csv")

In [16]:
pdfs_w_failed_pages

['100400',
 '100461',
 '100612',
 '101225',
 '101540',
 '101575',
 '102212',
 '102285',
 '102291']

In [17]:
len(unprocessed_pdfs - set(pdfs_w_failed_pages))

276

## Timeout pdfs

In [18]:
with open("/nrcan_p2/data/02_intermediate/20201006/geoscan/pdf/v1_20201125/timeout_files.txt", "r") as f:
    lines = f.readlines()

timeout_pdfs = []
for line in lines:
    if line.strip().endswith(".pdf"):
        pdf_id = re.match(r"/nrcan_p2/data/01_raw/20201006/geoscan/raw/pdf/(\d+)", line)[1]
        timeout_pdfs.append(pdf_id)

In [19]:
timeout_pdfs

['291805',
 '291808',
 '291809',
 '291810',
 '291811',
 '291812',
 '291900',
 '291901',
 '291920',
 '292021',
 '292074',
 '292100',
 '292101',
 '292187',
 '292214',
 '292279',
 '292306',
 '292316',
 '292378',
 '292379',
 '296624',
 '292483',
 '292503',
 '292551',
 '296674',
 '292553',
 '292555',
 '292559',
 '292564',
 '292579',
 '296826',
 '296831',
 '292588',
 '292593',
 '292616',
 '296834',
 '296885',
 '296912',
 '296916',
 '296944',
 '296945',
 '296952',
 '296995',
 '297045',
 '297208',
 '297225',
 '297236',
 '297336',
 '297455',
 '297512',
 '297515',
 '297608',
 '297641',
 '297643',
 '297645',
 '247749',
 '297648',
 '247750',
 '247751',
 '297668',
 '247752',
 '247753',
 '297720',
 '247754',
 '297747',
 '297776',
 '297820',
 '297821',
 '297864',
 '297891',
 '247758',
 '247834',
 '247853',
 '247867',
 '248213',
 '248226',
 '248231',
 '248252',
 '261482',
 '261487',
 '261488',
 '261500',
 '261570',
 '262739',
 '262740',
 '262741',
 '285377',
 '285402',
 '285475',
 '285518',
 '285524',

In [20]:
len(timeout_pdfs)

273

In [21]:
len(set(timeout_pdfs))

273

In [22]:
for pdf in timeout_pdfs:
    if pdf not in unprocessed_pdfs:
        print(pdf)

In [23]:
sorted(list(unprocessed_pdfs - set(timeout_pdfs)))

['211796', '247715', '326947']

211796: b3dc02bd-21a2-46ef-9120-08397e81ed3e TypeError: 'NoneType' object is not subscriptable
247715: empty file pdfminer returns empty object
326947: 71aa30bb-3896-4250-80c4-3340dd348b17 pdfminer.pdfdocument.PDFEncryptionError: Unsupported revision: param={'CF': {'StdCF': {'AuthEvent': /'DocOpen', 'CFM': /'AESV3', 'Length': 32}}, 'Filter': /'Standard', 'Length': 256, 'O': b'dy©u"HeƾjFQ«Àï@Éd?¶Áñ
nî ØÔäv#ý&vÄVÇ11L', 'OE': b'}ux÷øÁ>Þ]`!Î6Í&¡¬3 «r`2w4pÿ', 'P': -1052, 'Perms': b'8;q»Ëaå­ÓÅF?b', 'R': 6, 'StmF': /'StdCF', 'StrF': /'StdCF', 'U': b'`%®Ú×ÕüÄÅHÓM`âFG¡Ö}ãs!S(In°b-Ykw ÈxÅÿ', 'UE': b'.tQË:Þ_(çÚá½Ý2=b<ÞM=ö%°¥­', 'V': 5}

## Processed pdfs

In [7]:
frames = []
file_count = 0
path = "/nrcan_p2/data/02_intermediate/20201006/geoscan/pdf/v1_20201125"
for (dirpath, dirnames, filenames) in os.walk(path):
    for filename in filenames:
        if filename.endswith('.pdfminer_split.csv'):
            try:
                df_per_pdf = pd.read_csv(os.sep.join([dirpath, filename]), index_col=[0])
            except TypeError:
                print(filename)
                continue
            frames.append(df_per_pdf)
            file_count += 1
full_df = pd.concat(frames).reset_index(drop=True)

/opt/conda/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3145: DtypeWarning: Columns (7) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


299332.pdfminer_split.csv


MemoryError: Unable to allocate 988. MiB for an array with shape (4, 32370864) and data type float64